Laba_1_Kapustina

PolinaKapustina opened the pull request.

PolinaKapustina committed:
Add files via upload

PolinaKapustina committed:
Add files via upload

ArisovaAnastasia commented on src/tbitfield.cpp:
```diff
@@ -1,86 +1,196 @@
+	if (len >= 0)
+	{
+		BitLen = len;
+		MemLen = (len + (sizeof(TELEM) * 8 - 1)) >> sizeof(TELEM);
```
Формула неверна. Что она вычисляет?

PolinaKapustina replied:
Формула должна была вычислять значение MemLen, но в ней допущена ошибка. На данный момент она исправлена.

ArisovaAnastasia commented on src/tbitfield.cpp:
```diff
@@ -1,86 +1,196 @@
+{
+	if ((n > -1) || (n < BitLen))
+	return (pMem[GetMemIndex(n)] & GetMemMask(n)) == 0 ? 0 : 1;
+	else throw;
```
Все исключения должны быть с содержательным комментарием

PolinaKapustina replied:
Добавлен комментарий к исключению.

ArisovaAnastasia commented on src/tbitfield.cpp:
```diff
@@ -1,86 +1,196 @@
+
+int TBitField::operator==(const TBitField &bf) const // сравнение
+{
+	if (bf == *this)
```
Класс имеет сложную структуру, поэтому просто сравнивать указатели неправильно. Нужно реализовать поэлементное сравнение.

А при реализации оператора != можно будет использовать оператор ==

PolinaKapustina replied:
Спасибо. Исправила на прошлую полную реализацию этой части кода.

ArisovaAnastasia reviewed (commented):
Есть замечания. Некоторые тесты падают, это происходит в операторе ==

PolinaKapustina committed:
Update tbitfield.cpp

Замечания учтены, постаралась всё оперативно исправить

ArisovaAnastasia reviewed (commented):
Замечания исправлены верно. Но некоторые тесты падают, а именно:
TSet.can_insert_non_existing_element_using_plus_operator
TSet.can_insert_existing_element_using_plus_operator
TSet.can_intersect_two_sets_of_equal_size
TSet.can_intersect_two_sets_of_non_equal_size
TSet.check_negation_operator
TSet.can_combine_three_sets_of_non_equal_size

PolinaKapustina committed:
New tset.cpp

Думаю, что я исправила функциb из-за которых были ошибки, но корректность тестов не могу проверить так как при попытке запустить лабораторную на компьютере выдаёт ошибку "Нет данных", хотя все пути файлов соблюдены. Скачать файлы обратно с реквеста пока не вышло.

PolinaKapustina commented:
Предположительно нашла ошибки в функциях и исправила их. К сожалению на данный момент нет возможности проверить корректность тестов, так как студия выдаёт ошибку при попытке запустить лабораторную работу на компьютере.

ArisovaAnastasia commented on src/tset.cpp:
```diff
@@ -39,7 +39,8 @@ int TSet::GetMaxPower(void) const // получить макс. к-во эл-т
 int TSet::IsMember(const int Elem) const // элемент множества?
 {
-    return BitField.GetBit(Elem) == 0 ? 0 : 1;
+    if (Elem < 0 || Elem >= MaxPower) throw("The element is out of bounds of the set);
```
Программа не собирается, забыты закрывающие кавычки для комментария исключения.

ArisovaAnastasia commented on src/tset.cpp:
```diff
@@ -1,88 +1,164 @@
+}
+
+// конструктор преобразования типа
+TSet::TSet(const TBitField &bf) : BitField(-1)
```
Вы передаете в конструктор класса BitField len=-1. Реализация конструктора написана так, что при отрицательной длине выбрасывается исключение. Поэтому этот конструктор преобразования типа всегда будет выбрасывать исключение. Именно из-за этого не работали тесты. 
Необходимо вызывать конструктор копирования для BitField

ArisovaAnastasia commented on test/test_tset.cpp:
```diff
@@ -1,297 +1,325 @@
+	expSet.InsElem(3);
+	expSet.InsElem(4);
+	expSet.InsElem(6);
+		EXPECT_EQ(expSet, set3);
```
Тут опечатка ,наверно имелось в виду сравнение с set4 .

ArisovaAnastasia commented on test/test_tset.cpp:
```diff
@@ -1,297 +1,325 @@
+
+TEST(TSet, can_get_max_power_set)
+{
+  const int size = 5;
```
Во всем файле необходимо сделать отступы для кода в теле функций

ArisovaAnastasia commented on test/test_tbitfield.cpp:
```diff
@@ -1,311 +1,332 @@
+
+TEST(TBitField, can_create_bitfield_with_positive_length)
+{
+  ASSERT_NO_THROW(TBitField bf(3));
```
Во всем файле необходимо сделать отступы для кода в теле функций

ArisovaAnastasia reviewed (commented):
Еще раз посмотрела код, нашла почему падали тесты, это было из-за неправильной реализации конструктора преобразования типа. Все указанные замечания необходимо исправить

PolinaKapustina committed:
Update test_tset.cpp

PolinaKapustina committed:
Update tset.cpp

PolinaKapustina committed:
Update test_tbitfield.cpp

PolinaKapustina committed:
Update test_tbitfield.cpp

ArisovaAnastasia commented:
Лабораторная работа принята
